internal/cmd/gocdk/serve.go

```diff
@@ -16,10 +16,213 @@ package main
+	if err := buildForServe(ctx, pctx, moduleRoot, alloc.exePath); err != nil {
+		return err
+	}
+	process, err := alloc.start(ctx, pctx, logger)
+	if err != nil {
+		return err
+	}
+	logger.Printf("Serving at %s", alloc.url(""))
+	// TODO(light): Loop on filesystem change.
+	<-ctx.Done()
+	logger.Println("Shutting down...")
+	if err := signalGracefulShutdown(process.Process); err != nil {
+		process.Process.Kill()
+	}
+	process.Wait()
+	return nil
+}
+
+// buildForServe runs Wire and `go build` at moduleRoot to create exePath.
+func buildForServe(ctx context.Context, pctx *processContext, moduleRoot string, exePath string) error {
+	moduleEnv := append(append([]string(nil), pctx.env...), "GO111MODULE=on")
+
+	if wireExe, err := exec.LookPath("wire"); err == nil {
+		// TODO(light): Only run Wire if needed, but that requires source analysis.
+		wireCmd := exec.CommandContext(ctx, wireExe, "./...")
+		wireCmd.Dir = moduleRoot
+		wireCmd.Env = moduleEnv
+		// TODO(light): Collect build logs into error.
+		wireCmd.Stdout = pctx.stderr
+		wireCmd.Stderr = pctx.stderr
+		if err := wireCmd.Run(); err != nil {
+			return xerrors.Errorf("build server: wire: %w", err)
+		}
+	}
+
+	buildCmd := exec.CommandContext(ctx, "go", "build", "-o", exePath)
+	buildCmd.Dir = moduleRoot
+	buildCmd.Env = moduleEnv
+	// TODO(light): Collect build logs into error.
+	buildCmd.Stdout = pctx.stderr
+	buildCmd.Stderr = pctx.stderr
+	if err := buildCmd.Run(); err != nil {
+		return xerrors.Errorf("build server: go build: %w", err)
+	}
+
+	return nil
+}
+
+// serverAlloc stores the built executable path and port for a single instance
+// of the user's application server.
+type serverAlloc struct {
+	exePath string
+	port    int
+}
+
+// url returns the URL to connect to the server with.
+func (alloc *serverAlloc) url(path string) *url.URL {
+	if path == "" {
+		path = "/"
+	}
+	return &url.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("localhost:%d", alloc.port),
+		Path:   path,
+	}
+}
+
+// start starts the server process specified by the alloc and waits for
+// it to become healthy.
+func (alloc *serverAlloc) start(ctx context.Context, pctx *processContext, logger *log.Logger) (*exec.Cmd, error) {
+	// Run server.
+	logger.Print("Starting server...")
+	process := exec.Command(alloc.exePath, fmt.Sprintf("--address=localhost:%d", alloc.port))
+	process.Stdout = pctx.stdout
+	process.Stderr = pctx.stderr
+	if err := process.Start(); err != nil {
+		return nil, xerrors.Errorf("start server: %w", err)
+	}
+
+	// Server must report alive within 30 seconds.
+	aliveCtx, aliveCancel := context.WithTimeout(ctx, 30*time.Second)
+	err := waitForHealthy(aliveCtx, alloc.url("/healthz/liveness"))
+	aliveCancel()
+	if err != nil {
+		process.Process.Kill() // Send SIGKILL; no graceful shutdown.
+		process.Wait()
+		return nil, xerrors.Errorf("start server: %w", err)
+	}
+
+	// Wait for server to be ready.
+	logger.Printf("Waiting for server %s to report ready...", alloc.url("/"))
```

Comment:
Add one of these message above for waiting for live?

Reply:
fwiw, this is what all `rails s` outputs when it starts up.  [RailsGuides source](https://guides.rubyonrails.org/command_line.html#rails-server)

<img width="641" alt="Screen Shot 2019-04-16 at 11 36 14 AM" src="https://user-images.githubusercontent.com/4240737/56235329-1352ab80-603c-11e9-8325-3c45fc5d7795.png">

Reply:
And for comparison, with this PR `gocdk serve` outputs:

```
$ gocdk serve
gocdk: 2019/04/16 12:20:46 Preparing to serve dev...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
gocdk: 2019/04/16 12:20:46 Building server...
gocdk: 2019/04/16 12:20:49 Starting server...
gocdk: 2019/04/16 12:20:49 Waiting for server http://localhost:8080/ to report ready...
gocdk: 2019/04/16 12:20:49 Serving at http://localhost:8080/
```

-  I like the Ctrl-C message; I'll add that one in.
-  Once we have some CLI version info that would likely be good to include that.
-  We probably want to condense the Terraform output, since it's a bit overpowering. #1821 would address that.

Reply:
LGTM. +1 on hiding the Terraform stuff. It's particularly odd for local serving.